internal/pkg/skuba/replica/replica.go

```diff
@@ -0,0 +1,409 @@
+	replicaJSON := fmt.Sprintf(patchReplicas, size)
+	return r.Client.AppsV1().Deployments(metav1.NamespaceSystem).Patch(r.Deployment.ObjectMeta.Name, types.StrategicMergePatchType, []byte(replicaJSON))
+}
+
+// UpdateNodes patches for replicas affinity rules after adding of nodes.
+// This updates affinity rule to preferredDuringSchedulingIgnoredDuringExecution of
+// cluster size less then replica size,
+// And updates affinity rule to requiredDuringSchedulingIgnoredDuringExecution when
+// cluster size is equal or greater to replicas.
+func (r *Helper) UpdateNodes() error {
+	if err := r.deploymentsHelper(); err != nil {
+		return err
+	}
+
+	for _, deployment := range r.Deployments.Items {
+		if err := r.deploymentHelper(deployment.Name); err != nil {
+			return err
+		}
+		r.ReplicaSize = int(*r.Deployment.Spec.Replicas)
+		if r.ReplicaSize == 0 {
+			r.ReplicaSize = r.MinSize
+		}
+
+		nodes := r.ClusterSize
+		updated := false
+		var err error
+		switch {
+		case nodes >= r.ReplicaSize:
+			updated, err = r.replaceAffinity(preferred, required)
+			if err != nil {
+				return err
+			}
+		case nodes >= r.MinSize:
+			updated, err = r.replaceAffinity(required, preferred)
+			if err != nil {
+				return err
+			}
+		}
+		if !updated && nodes <= r.ReplicaSize {
+			// update replicas to trigger pod re-distribution when affinity is in
+			// preferredDuringSchedulingIgnoredDuringExecution.
+			if _, err := r.updateDeploymentReplica(r.ClusterSize - 1); err != nil {
+				return err
+			}
+			if _, err := r.updateDeploymentReplica(r.ReplicaSize); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// UpdateBeforeNodeDrains patches affinity rules and wait for pod running before node remove.
+// Patch will apply to deployment with preferredDuringSchedulingIgnoredDuringExecution affinity
+// when cluster size equal or less then replicas.
+func (r *Helper) UpdateBeforeNodeDrains() error {
+	if err := r.deploymentsHelper(); err != nil {
+		return err
+	}
+
+	for _, deployment := range r.Deployments.Items {
+		if err := r.deploymentHelper(deployment.Name); err != nil {
+			return err
+		}
+		r.ReplicaSize = int(*r.Deployment.Spec.Replicas)
+		if r.ClusterSize > r.ReplicaSize {
+			continue
+		}
+
+		updated, err := r.replaceAffinity(required, preferred)
+		if err != nil {
+			return err
+		}
+		if !updated && r.ClusterSize <= r.ReplicaSize {
+			// update replicas to trigger pod re-distribution when affinity is in
+			// preferredDuringSchedulingIgnoredDuringExecution.
+			if _, err := r.updateDeploymentReplica(r.ClusterSize - 1); err != nil {
+				return err
+			}
+			if _, err := r.updateDeploymentReplica(r.ReplicaSize); err != nil {
+				return err
+			}
+			updated = true
+		}
+		if updated {
+			var err error
+			retry := retryCount
+			err = r.waitForUpdates(&retry, &err)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *Helper) waitForUpdates(retry *int, e *error) error {
```

Comment:
```suggestion
func (r *Helper) waitForUpdates(retry int) error {
```

Reply:
e is to return the last error when retry exhausted. And the pointer is there so the function can update the e instead of creating new e memory block for every recursive call. ... This is how I understand, but I might be wrong.

Reply:
Still I didn’t get the point because at the end u still just need one (last) error except you would like aggregate all errors but it seems not worthy in this case.

I don’t think a need to reserve one variable space for reusing the same error variable space. Suggest just use general pattern by returning error then checking that should be good enough. WDYT.

Reply:
I am fine with that.

Reply:
pointers removed